File: examples/usb-xmega-1.0/ucecho/avr/ucecho.c
/*!
   ucecho -- uppercase conversion example for all EZ-USB devices
   Copyright (C) 2009-2010 ZTEX e.K.
   http://www.ztex.de

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 3 as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, see http://www.gnu.org/licenses/.
!*/

#include <avr/io.h>

#define F_CPU 32000000UL
#include <util/delay.h>

typedef uint8_t byte;

int main(void)	
{
    // enable 32.768 kHz, 32 MHz, 2 Mhz clocks
    asm volatile (
	"ldi r24,0xd8" "\n\t"
	"ldi r23,7" "\n\t"
	"out 0x34,r24" "\n\t"
	"sts 0x50,r23"
	::: "r24", "r23" 
    ); 
    
    // wait until clocks are ready
    while ( (OSC.STATUS & 7) != 7 ) { }

    // enable run time configuration of 32 MHz and 2 MHz clocks; select 32 MHz clock as system clock
    asm volatile (
	"ldi r24,0xd8" "\n\t"
	"ldi r23,1" "\n\t"
	"out 0x34,r24" "\n\t"
	"sts 0x60,r23" "\n\t"
	"out 0x34,r24" "\n\t"
	"sts 0x68,r23" "\n\t"
	"out 0x34,r24" "\n\t"
	"sts 0x40,r23"
	::: "r24", "r23" 
    ); 

    // disable JTAG at portb
    asm volatile (
	"ldi r24,0xd8" "\n\t"
	"ldi r23,1" "\n\t"
	"out 0x34,r24" "\n\t"
	"sts 0x96,r23" "\n\t"
	::: "r24", "r23" 
    ); 
     
    
    PORTH.DIR = 0;	// XMEGA port H (= EZ-USB PORT D) is input
    PORTK.DIR = 255;	// XMEGA port K (= EZ-USB PORT B) is output
    
    while (1) {	

	byte b = PORTH.IN;
	PORTK.OUT = ( b >= 'a' && b <= 'z' ) ? b - 32 : b;

    }
}

File: examples/usb-fpga-1.2/ucecho/UCEcho.java
/*!
   ucecho -- uppercase conversion example for ZTEX USB-FPGA Module 1.2
   Copyright (C) 2009-2011 ZTEX GmbH.
   http://www.ztex.de

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 3 as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, see http://www.gnu.org/licenses/.
!*/

import java.io.*;
import java.util.*;

import ch.ntb.usb.*;

import ztex.*;

// *****************************************************************************
// ******* ParameterException **************************************************
// *****************************************************************************
// Exception the prints a help message
class ParameterException extends Exception {
    public final static String helpMsg = new String (
		"Parameters:\n"+
		"    -d <number>       Device Number (default: 0)\n" +
		"    -f 	       Force uploads\n" +
		"    -p                Print bus info\n" +
		"    -w                Enable certain workarounds\n"+
		"    -h                This help" );
    
    public ParameterException (String msg) {
	super( msg + "\n" + helpMsg );
    }
}

// *****************************************************************************
// ******* Test0 ***************************************************************
// *****************************************************************************
class UCEcho extends Ztex1v1 {

// ******* UCEcho **************************************************************
// constructor
    public UCEcho ( ZtexDevice1 pDev ) throws UsbException {
	super ( pDev );
    }

// ******* echo ****************************************************************
// writes a string to Endpoint 4, reads it back from Endpoint 2 and writes the output to System.out
    public void echo ( String input ) throws UsbException {
	byte buf[] = input.getBytes(); 
	int i = LibusbJava.usb_bulk_write(handle(), 0x04, buf, buf.length, 1000);
	if ( i<0 )
	    throw new UsbException("Error sending data: " + LibusbJava.usb_strerror());
	System.out.println("Send "+i+" bytes: `"+input+"'" );

	try {
    	    Thread.sleep( 10 );
	}
	    catch ( InterruptedException e ) {
	}

	buf = new byte[1024];
	i = LibusbJava.usb_bulk_read(handle(), 0x82, buf, 1024, 1000);
	if ( i<0 )
	    throw new UsbException("Error receiving data: " + LibusbJava.usb_strerror());
	System.out.println("Read "+i+" bytes: `"+new String(buf,0,i)+"'" );  
    }
    
// ******* main ****************************************************************
    public static void main (String args[]) {
    
	int devNum = 0;
	boolean force = false;
	boolean workarounds = false;
	
	try {
// init USB stuff
	    LibusbJava.usb_init();

// scan the USB bus
	    ZtexScanBus1 bus = new ZtexScanBus1( ZtexDevice1.ztexVendorId, ZtexDevice1.ztexProductId, true, false, 1);
	    if ( bus.numberOfDevices() <= 0) {
		System.err.println("No devices found");
	        System.exit(0);
	    }
	    
// scan the command line arguments
    	    for (int i=0; i<args.length; i++ ) {
	        if ( args[i].equals("-d") ) {
	    	    i++;
		    try {
			if (i>=args.length) throw new Exception();
    			devNum = Integer.parseInt( args[i] );
		    } 
		    catch (Exception e) {
		        throw new ParameterException("Device number expected after -d");
		    }
		}
		else if ( args[i].equals("-f") ) {
		    force = true;
		}
		else if ( args[i].equals("-p") ) {
	    	    bus.printBus(System.out);
		    System.exit(0);
		}
		else if ( args[i].equals("-p") ) {
	    	    bus.printBus(System.out);
		    System.exit(0);
		}
		else if ( args[i].equals("-w") ) {
	    	    workarounds = true;
		}
		else if ( args[i].equals("-h") ) {
		        System.err.println(ParameterException.helpMsg);
	    	        System.exit(0);
		}
		else throw new ParameterException("Invalid Parameter: "+args[i]);
	    }
	    

// create the main class	    
	    UCEcho ztex = new UCEcho ( bus.device(devNum) );
	    ztex.certainWorkarounds = workarounds;
	    
// upload the firmware if necessary
	    if ( force || ! ztex.valid() || ! ztex.dev().productString().equals("ucecho example for UFM 1.2")  ) {
		System.out.println("Firmware upload time: " + ztex.uploadFirmware( "ucecho.ihx", force ) + " ms");
	    }
	    
// upload the bitstream if necessary
	    if ( force || ! ztex.getFpgaConfiguration() ) {
		System.out.println("FPGA configuration time: " + ztex.configureFpga( "fpga/ucecho.bit" , force ) + " ms");
	    } 


// claim interface 0
	    ztex.trySetConfiguration ( 1 );
	    ztex.claimInterface ( 0 );
	    
// read string from stdin and write it to USB device
	    String str = "";
	    BufferedReader reader = new BufferedReader( new InputStreamReader( System.in ) );
	    while ( ! str.equals("quit") ) {
		System.out.print("Enter a string or `quit' to exit the program: ");
		str = reader.readLine();
		if ( ! str.equals("") )
		    ztex.echo(str);
	        System.out.println("");
	    }
	    
// release interface 0
	    ztex.releaseInterface( 0 );	
	    
	}
	catch (Exception e) {
	    System.out.println("Error: "+e.getLocalizedMessage() );
	} 
   } 
   
}
